### ISSUE TITLE
`debug_traceblockByNumber` works, while explicitly tracing transaction from the block via `traceTransaction` fails

### ISSUE THREAD

**Issue by alexqrid** [binance-chain/bsc#373]
#### System information

Geth version: [`bsc-geth`](https://github.com/binance-chain/bsc/releases/tag/v1.1.1)
OS & Version:`Linux Debian Buster 5.11.0-051100-generic #202102142330 SMP Sun Feb 14 23:33:21 UTC 2021 x86_64 GNU/Linux`
#### Expected behaviour
I have a synchronizing full archive node, that is almost synced. While tracing (`debug_traceBlockByNumber`) a block number `7237611` using js `callTracer`, I receive a correct trace for all transactions in the block. When I try to trace the latest transaction in the block (`debug_traceTransaction`) I expect to see the same trace for the transaction as in response from `debug_traceBlockByNumber`.

#### Actual behaviour

<details><summary> Block 7237611 trace</summary>

Query

```json
{
    "jsonrpc": "2.0",
    "method": "debug_traceBlockByNumber",
    "params": ["0x6E6FEB",{"tracer":"callTracer"}],
    "id": 1
}
```

Response (included only last 2 transactions that have a problem with)
```json
....
        {
            "result": {
                "type": "CALL",
                "from": "0x35e7a025f4da968de7e4d7e4004197917f4070f1",
                "to": "0x0000000000000000000000000000000000001002",
                "value": "0x3504ef16e863db",
                "gas": "0x7fffffffffffadf7",
                "gasUsed": "0x5ec",
                "input": "0x",
                "output": "0x",
                "time": "22.76µs"
            }
        },
        {
            "result": {
                "type": "CALL",
                "from": "0x35e7a025f4da968de7e4d7e4004197917f4070f1",
                "to": "0x0000000000000000000000000000000000001000",
                "value": "0x31b4a02579dd9d8",
                "gas": "0x7fffffffffffac53",
                "gasUsed": "0x42fe",
                "input": "0xf340fa0100000000000000000000000035e7a025f4da968de7e4d7e4004197917f4070f1",
                "output": "0x",
                "time": "3.794206ms"
            }
        }
```
</details>

`debug_traceTransaction` for the last transaction `0x91f8f8056af4242208a3eed09aaa3cf8a4a78b63ff6bdae503149dfbd0d68acf` in the block `7237611` fails with the following error:  
<details><summary> API response</summary>

Query
```json
{
    "jsonrpc": "2.0",
    "method": "debug_traceTransaction",
    "params": ["0x91f8f8056af4242208a3eed09aaa3cf8a4a78b63ff6bdae503149dfbd0d68acf",{"tracer":"callTracer"}],
    "id": 1
}
```

Response

```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "error": {
        "code": -32000,
        "message": "transaction 0x0e04d8d28d42bd0f6bd9bd5ee27b88053043711ad43b97687bc85a1a197a8b47 failed: insufficient funds for transfer: address 0x35E7a025f4da968De7e4D7E4004197917F4070F1"
    }
}
```

<summary> Trace of 0x0e04d8d28d42bd0f6bd9bd5ee27b88053043711ad43b97687bc85a1a197a8b47</summary>

```json
{
    "jsonrpc": "2.0",
    "method": "debug_traceTransaction",
    "params": ["0x0e04d8d28d42bd0f6bd9bd5ee27b88053043711ad43b97687bc85a1a197a8b47",{"tracer":"callTracer"}],
    "id": 1
}
```

Response

```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
        "type": "CALL",
        "from": "0x35e7a025f4da968de7e4d7e4004197917f4070f1",
        "to": "0x0000000000000000000000000000000000001002",
        "value": "0x3504ef16e863db",
        "gas": "0x7fffffffffffadf7",
        "gasUsed": "0x5ec",
        "input": "0x",
        "output": "0x",
        "time": "17.03µs"
    }
}
```
</details>

#### Steps to reproduce the behaviour
Query the api to get the trace of transaction `0x91f8f8056af4242208a3eed09aaa3cf8a4a78b63ff6bdae503149dfbd0d68acf`

#### The problem analysis
Actually the error occurs while getting the [`StateAtTransaction`](https://github.com/binance-chain/bsc/blob/03f7b318d9e19d0477bf3ffe6d6f85a681048dae/eth/tracers/api.go#L730) in the [`state_accessor.go`](https://github.com/binance-chain/bsc/blob/03f7b318d9e19d0477bf3ffe6d6f85a681048dae/eth/state_accessor.go#L175) because the execution of the previous transaction (`0x0e04d8d28d42bd0f6bd9bd5ee27b88053043711ad43b97687bc85a1a197a8b47`) fails, as you can see from the api response above.  

But the transaction receipts of these transactions have a successful status, i.e., they are not actually failed transactions.
<details><summary> Receipt of 0x91f8f8056af4242208a3eed09aaa3cf8a4a78b63ff6bdae503149dfbd0d68acf</summary>

```json
                "root": "0x",
                "status": "0x1",
                "cumulativeGasUsed": "0x289075e",
                "logsBloom": "0x00004000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002010000000000000000000000000000000000020000200000000000080000000000000000000000000000000000000000000000000000000000000000000000000000400000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000",
                "logs": [
                    {
                        "address": "0x0000000000000000000000000000000000001000",
                        "topics": [
                            "0x93a090ecc682c002995fad3c85b30c5651d7fd29b0be5da9d784a3302aedc055",
                            "0x00000000000000000000000035e7a025f4da968de7e4d7e4004197917f4070f1"
                        ],
                        "data": "0x000000000000000000000000000000000000000000000000031b4a02579dd9d8",
                        "blockNumber": "0x6e6feb",
                        "transactionHash": "0x91f8f8056af4242208a3eed09aaa3cf8a4a78b63ff6bdae503149dfbd0d68acf",
                        "transactionIndex": "0x207",
                        "blockHash": "0x24c52f0091df84bcff7090ef197db56140b29c6ec6c8c4204fde4f2ee7e2d626",
                        "logIndex": "0x2eb",
                        "removed": false
                    }
                ],
                "transactionHash": "0x91f8f8056af4242208a3eed09aaa3cf8a4a78b63ff6bdae503149dfbd0d68acf",
                "contractAddress": "0x0000000000000000000000000000000000000000",
                "gasUsed": "0x42fe",
                "blockHash": "0x24c52f0091df84bcff7090ef197db56140b29c6ec6c8c4204fde4f2ee7e2d626",
                "blockNumber": "0x6e6feb",
                "transactionIndex": "0x207"
```
</details>

<details><summary> Receipt of 0x0e04d8d28d42bd0f6bd9bd5ee27b88053043711ad43b97687bc85a1a197a8b47</summary>

```json
                "root": "0x",
                "status": "0x1",
                "cumulativeGasUsed": "0x288c460",
                "logsBloom": "0x00004000000000000000004000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000080000000000000000000080000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000104000000000000000000000000000000000000000010000000000000000000000000000000000",
                "logs": [
                    {
                        "address": "0x0000000000000000000000000000000000001002",
                        "topics": [
                            "0x6c98249d85d88c3753a04a22230f595e4dc8d3dc86c34af35deeeedc861b89db",
                            "0x00000000000000000000000035e7a025f4da968de7e4d7e4004197917f4070f1"
                        ],
                        "data": "0x000000000000000000000000000000000000000000000000003504ef16e863db",
                        "blockNumber": "0x6e6feb",
                        "transactionHash": "0x0e04d8d28d42bd0f6bd9bd5ee27b88053043711ad43b97687bc85a1a197a8b47",
                        "transactionIndex": "0x206",
                        "blockHash": "0x24c52f0091df84bcff7090ef197db56140b29c6ec6c8c4204fde4f2ee7e2d626",
                        "logIndex": "0x2ea",
                        "removed": false
                    }
                ],
                "transactionHash": "0x0e04d8d28d42bd0f6bd9bd5ee27b88053043711ad43b97687bc85a1a197a8b47",
                "contractAddress": "0x0000000000000000000000000000000000000000",
                "gasUsed": "0x5ec",
                "blockHash": "0x24c52f0091df84bcff7090ef197db56140b29c6ec6c8c4204fde4f2ee7e2d626",
                "blockNumber": "0x6e6feb",
                "transactionIndex": "0x206"
```
</details>

Unfortunately this is not the only case. I found another case, when the block trace fails (`7379228`) with the  error  
`"insufficient funds for transfer: address 0xd6caA02BBebaEbB5d7e581e4B66559e635F805fF"`. It was the last transaction of the block number `7379228` with the hash `0x3465af5b898cbdab9566129a951bfdf31807c1c46496e29f8959ccbdab445ba3`

<details><summary>Trace of 0x3465af5b898cbdab9566129a951bfdf31807c1c46496e29f8959ccbdab445ba3</summary>
Query

```json
{
    "jsonrpc": "2.0",
    "method": "debug_traceTransaction",
    "params": ["0x3465af5b898cbdab9566129a951bfdf31807c1c46496e29f8959ccbdab445ba3",{"tracer":"callTracer"}],
    "id": 1
}
```

Response

```json

{
    "jsonrpc": "2.0",
    "id": 1,
    "error": {
        "code": -32000,
        "message": "transaction 0x72a02abfff3e85071b9c4dc1401de538899fc94b512405ec690b94246f97fafc failed: insufficient funds for transfer: address 0xd6caA02BBebaEbB5d7e581e4B66559e635F805fF"
    }
}
```
Query

```json
{
    "jsonrpc": "2.0",
    "method": "debug_traceTransaction",
    "params": ["0x72a02abfff3e85071b9c4dc1401de538899fc94b512405ec690b94246f97fafc",{"tracer":"callTracer"}],
    "id": 1
}
```

Response

```json
  "type": "CALL",
  "from": "0xd6caa02bbebaebb5d7e581e4b66559e635f805ff",
  "to": "0x0000000000000000000000000000000000001002",
  "value": "0x68eab1520fcf73",
  "gas": "0x7fffffffffffadf7",
  "gasUsed": "0x5ec",
  "input": "0x",
  "output": "0x",
  "time": "16.67µs"
```
</details>

But if you try to trace the transaction, that was the reason of fail (`0x72a02abfff3e85071b9c4dc1401de538899fc94b512405ec690b94246f97fafc`), explicitly with `debug_traceTransaction` it succeeds.
Moreover this case can be met rather more than the first case. 
#### Summary
Overall the problem with tracing transactions seems to be present  in last 2 transactions of some blocks, i.e., with the transactions that miner sends to the system contracts. May be this is somehow related. 
I believe the behavior should be the same for all cases: if block tracing fails because of any transaction then the explicit trace of the transaction, that failed while tracing the block, should not succeed.

**Comment by maxogden** [comment]
I have the same problem in https://github.com/binance-chain/bsc/issues/314 with debug_standardTraceBlockToFile. If one of the txs in the block fails due to the insufficient funds for transfer error it stops processing the rest of the txs in the block

**Comment by alexqrid** [comment]
@maxogden I have made a PR that resolves the issue #374

**Comment by guagualvcha** [comment]
fixed in https://github.com/binance-chain/bsc/pull/374
